[mlir][spirv] Fix lookup logic `spirv.target_env` for `gpu.module`

oojahooo opened the pull request:
The `gpu.module` operation can contain `spirv.target_env` attributes within an array attribute named `"targets"`. So it accounts for that case by iterating over the `"targets"` attribute, if present, and looking up `spirv.target_env`.

oojahooo committed:
[mlir][SPIRV] Fix lookup logic `spirv.target_env` for `gpu.module`

The `gpu.module` operation can contain `spirv.target_env` attributes
within an array attribute named `"targets"`. So it accounts for that
case by iterating over the `"targets"` attribute, if present, and
looking up `spirv.target_env`.

IgWod-IMG reviewed (commented):
A rather naive question, as I'm not familiar with the GPU dialect. Is it possible for the target env to be present in both `"targets"` and as an op attribute at the same time? If so, does the look-up order matter? I’d hope the same target env is returned either way, but hypothetically if there is a different env in `"targets"` and in the op, I wonder which one should take a precedence?

Also, it'd be nice to have a test for the change. Probably in GPU to SPIR-V conversion test directory?

kuhar reviewed (changes requested):
Could you add a test that shows where this comes up? I'm not sure if we should look up the target env in`"targets"` attribute in **any** of possible parents

oojahooo committed:
Add test

oojahooo committed:
Delete braces for complying with conding standard

It resolves
https://github.com/llvm/llvm-project/pull/147262#discussion_r2189731690

oojahooo commented on mlir/lib/Dialect/SPIRV/IR/TargetAndABI.cpp:
```diff
@@ -184,6 +184,15 @@ spirv::TargetEnvAttr spirv::lookupTargetEnv(Operation *op) {
     if (!op)
       break;
 
+    if (auto arrAttr = op->getAttrOfType<ArrayAttr>("targets")) {
```
Thank you for your information. I committed.

oojahooo commented:
> A rather naive question, as I'm not familiar with the GPU dialect. Is it possible for the target env to be present in both `"targets"` and as an op attribute at the same time? If so, does the look-up order matter? I’d hope the same target env is returned either way, but hypothetically if there is a different env in `"targets"` and in the op, I wonder which one should take a precedence?
> 
> Also, it'd be nice to have a test for the change. Probably in GPU to SPIR-V conversion test directory?

When the `gpu.module` op already has a target env as a dict-attr, running the `spirv-attach-target` pass can reproduce the situation you described. In such cases, I believe that the target env inserted by `spirv-attach-target` can be occurred without external intervention, so it should be given the highest priority during lookup. If that's the case, I think the flow in the committed code would be appropriate.

kuhar commented:
> If target env is inserted via the `spirv-attach-target` pass when no op—including `builtin.module` or `gpu.module`—has an existing one, then under the current implementation, the target env is attached only to the `"targets"` array attribute of `gpu.module`, and does not appear elsewhere.
> When the `convert-gpu-to-spirv` pass is run in this case, it fails to find the target env in `"targets"` and instead falls back to the default target env, which leads to a legalization failure of the `memref.load` op.

To me the sounds like the wrong fix. Could we teach the passes you mentioned to also look in `"targets"` instead of relying on the core `lookupTargetEnv` helper that doesn't know anything about the `gpu` dialect?

oojahooo commented:
> > If target env is inserted via the `spirv-attach-target` pass when no op—including `builtin.module` or `gpu.module`—has an existing one, then under the current implementation, the target env is attached only to the `"targets"` array attribute of `gpu.module`, and does not appear elsewhere.
> > When the `convert-gpu-to-spirv` pass is run in this case, it fails to find the target env in `"targets"` and instead falls back to the default target env, which leads to a legalization failure of the `memref.load` op.
> 
> To me the sounds like the wrong fix. Could we teach the passes you mentioned to also look in `"targets"` instead of relying on the core `lookupTargetEnv` helper that doesn't know anything about the `gpu` dialect?

If I understand your point correctly, rather than hardcoding the `lookupTargetEnv` helper to directly search for the `"targets"` attribute, which is specific to the `gpu` dialect, it would be more appropriate to handle this logic within the `convert-gpu-to-spirv` pass.

There are two possible approaches that I think:

1. As you suggested, we could modify the `convert-gpu-to-spirv` pass to explicitly inspect the `"targets"` attribute of `gpu.module` and retrieve the corresponding `spirv.target_env`. If that lookup fails, we can fall back to invoking the `spirv::lookupTargetEnvOrDefault` function. I believe this is a reasonable solution since the pass is already coupled with the `gpu` dialect by design.

2. Alternatively, we could generalize the logic of `lookupTargetEnv` to retrieve all attributes via `getAttrDictionary()` and recursively walk through them to locate a `spirv.target_env` attribute. While this approach differs slightly from your suggestion, I thought it could more robustly support the intended semantics of passes using `lookupTargetEnvOrDefault`, and would also remain dialect-agnostic.

I would greatly appreciate it if you could confirm whether this interpretation is aligned with your suggestion. If so, I would be happy to revise the implementation accordingly based on the preferred direction.

Thank you for your support.

kuhar commented:
Approach 1. sounds to me like the way to go

oojahooo committed:
fixup! [mlir][SPIRV] Fix lookup logic `spirv.target_env` for `gpu.module`

Add lookup target env in "targets" attr logic to GPUToSPIRV pass

oojahooo committed:
Remove unnecessary check in test

oojahooo committed:
Add same lookup target env logic in GPUModuleConversion

oojahooo commented:
> Approach 1. sounds to me like the way to go

Thank you for replying.
I reverted logic of `lookupTargetEnv` and added new helper function named `lookupTargetEnvInTargets`, which finds spirv target env in `"targets"`, into GPUToSPIRV code.
Also I added same lookup logic in `GPUModuleConversion` for remaining target env during conversion.

oojahooo commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -48,19 +48,38 @@ struct GPUToSPIRVPass final : impl::ConvertGPUToSPIRVBase<GPUToSPIRVPass> {
   bool mapMemorySpace;
 };
 
+spirv::TargetEnvAttr
```
Thank you for your comment.

Since I intentionally did not include `OrDefault` in the function name `GPUToSPIRVPass::lookupTargetEnvInTargets`, I wanted the function to only attempt to retrieve the target environment from the `"targets"` attribute, reflecting that limited scope directly in the name.

Then, I defined a separate `GPUToSPIRVPass::lookupTargetEnvOrDefault` function that calls it and falls back to `spirv::lookupTargetEnvOrDefault` if needed, making the fallback behavior more explicit.

However, if adding multiple new functions is considered undesirable, I'm happy to simplify the implementation by having only `GPUToSPIRVPass::lookupTargetEnvOrDefault` and letting it handle whole condition.

IgWod-IMG replied:
> However, if adding multiple new functions is considered undesirable, I'm happy to simplify the implementation by having only `GPUToSPIRVPass::lookupTargetEnvOrDefault` and letting it handle whole condition.

I don't think it's undesirable. I just felt like removing the extra indirection may offer a nicer solution, but you made some good argument about why it makes sense the way it is. Initially I thought that you could add `OrDefault` in the name of the first function, but now I think it may be more confusing. Let's keep it the way it is, unless someone else has a strong opinion about it.

oojahooo replied:
Thank you for the thoughtful suggestion and for being open to my perspective!

oojahooo committed:
fixup! Add same lookup target env logic in GPUModuleConversion

remove unnecessary namespace qualifiers

oojahooo commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -385,6 +385,10 @@ LogicalResult GPUModuleConversion::matchAndRewrite(
     spvModule->setAttr(spirv::getTargetEnvAttrName(), attr);
+  for (auto targetAttr : moduleOp.getTargetsAttr())
+    if (auto spirvTargetEnvAttr =
+            llvm::dyn_cast<spirv::TargetEnvAttr>(targetAttr))
```
Thanks! I edited it.

kuhar replied:
This hasn't been resolved

oojahooo commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -86,7 +105,7 @@ void GPUToSPIRVPass::runOnOperation() {
   for (Operation *gpuModule : gpuModules) {
     spirv::TargetEnvAttr targetAttr =
-        spirv::lookupTargetEnvOrDefault(gpuModule);
+        lookupTargetEnvOrDefault(llvm::cast<gpu::GPUModuleOp>(gpuModule));
```
Thanks! I edited it.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -48,19 +48,38 @@ struct GPUToSPIRVPass final : impl::ConvertGPUToSPIRVBase<GPUToSPIRVPass> {
 
+spirv::TargetEnvAttr
+GPUToSPIRVPass::lookupTargetEnvInTargets(gpu::GPUModuleOp moduleOp) {
+  for (auto &targetAttr : moduleOp.getTargetsAttr())
```
Use spell out the type here since it's not obvious without an IDE/lsp: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

oojahooo replied:
Thanks! I edited it.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -48,19 +48,38 @@ struct GPUToSPIRVPass final : impl::ConvertGPUToSPIRVBase<GPUToSPIRVPass> {
+
+spirv::TargetEnvAttr
+GPUToSPIRVPass::lookupTargetEnvOrDefault(gpu::GPUModuleOp moduleOp) {
+  if (auto targetEnvAttr = lookupTargetEnvInTargets(moduleOp))
```
also here

oojahooo replied:
Thanks! I edited it.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -48,19 +48,37 @@ struct GPUToSPIRVPass final : impl::ConvertGPUToSPIRVBase<GPUToSPIRVPass> {
   void runOnOperation() override;
 
 private:
+  spirv::TargetEnvAttr lookupTargetEnvInTargets(gpu::GPUModuleOp moduleOp);
+  spirv::TargetEnvAttr lookupTargetEnvOrDefault(gpu::GPUModuleOp moduleOp);
```
Can you add documentation comments explaining what these do?

oojahooo replied:
Thanks! I added it.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -385,6 +385,9 @@ LogicalResult GPUModuleConversion::matchAndRewrite(
   if (auto attr = moduleOp->getAttrOfType<spirv::TargetEnvAttr>(
           spirv::getTargetEnvAttrName()))
     spvModule->setAttr(spirv::getTargetEnvAttrName(), attr);
+  for (auto targetAttr : moduleOp.getTargetsAttr())
```
Use the actual type here, it's not obvious outside of an IDE: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

oojahooo replied:
Thanks! I edited it.

kuhar commented on mlir/test/Conversion/GPUToSPIRV/lookup-target-env.mlir:
```diff
@@ -0,0 +1,17 @@
+// RUN: mlir-opt --spirv-attach-target='caps=Shader exts=SPV_KHR_storage_buffer_storage_class' --convert-gpu-to-spirv %s -o - | FileCheck %s
```
```suggestion
// RUN: mlir-opt --spirv-attach-target='caps=Shader exts=SPV_KHR_storage_buffer_storage_class' --convert-gpu-to-spirv %s | FileCheck %s
```

kuhar replied:
What do you think about running `convert-gpu-to-spirv` only? In the current form, it's not very clear where the 'targets' attribute is materialized.

oojahooo committed:
Update mlir/test/Conversion/GPUToSPIRV/lookup-target-env.mlir

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>

oojahooo committed:
Spell out the not obvious types

oojahooo committed:
Add documentation comments

oojahooo committed:
Simplify pass pipeline of test code

oojahooo committed:
Fix failures in other tests

It had to be checked for a lack of `"targets"` attributes in a GPU
module

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -385,9 +385,11 @@ LogicalResult GPUModuleConversion::matchAndRewrite(
-  for (const Attribute &targetAttr : moduleOp.getTargetsAttr())
-    if (auto spirvTargetEnvAttr = dyn_cast<spirv::TargetEnvAttr>(targetAttr))
-      spvModule->setAttr(spirv::getTargetEnvAttrName(), spirvTargetEnvAttr);
+  if (const ArrayAttr &targets = moduleOp.getTargetsAttr()) {
+    for (const Attribute &targetAttr : targets)
```
I think we can use values here instead of references, MLIR IR constructs are all designed to be passed by value

oojahooo replied:
Oh, I see. Thank you for the comment. I commited.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -61,9 +61,11 @@ struct GPUToSPIRVPass final : impl::ConvertGPUToSPIRVBase<GPUToSPIRVPass> {
-  for (const Attribute &targetAttr : moduleOp.getTargetsAttr())
-    if (auto spirvTargetEnvAttr = dyn_cast<spirv::TargetEnvAttr>(targetAttr))
-      return spirvTargetEnvAttr;
+  if (const ArrayAttr &targets = moduleOp.getTargetsAttr()) {
```
also here

oojahooo committed:
Do not use reference, do pass by value when assign attributes

oojahooo commented:
Hi @kuhar,
I think I've addressed all your previous feedback.
Let me know if there’s anything more you’d like to see changed. Otherwise, I’d appreciate your approval when you get a chance. Thanks!

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -385,6 +385,11 @@ LogicalResult GPUModuleConversion::matchAndRewrite(
     spvModule->setAttr(spirv::getTargetEnvAttrName(), attr);
+  if (ArrayAttr targets = moduleOp.getTargetsAttr()) {
+    for (Attribute targetAttr : targets)
+      if (auto spirvTargetEnvAttr = dyn_cast<spirv::TargetEnvAttr>(targetAttr))
+        spvModule->setAttr(spirv::getTargetEnvAttrName(), spirvTargetEnvAttr);
```
What happens when there's more than one target env attribute in the array? IIUC, this will always pick the last one.

oojahooo replied:
As I mentioned in the comment above, I’ve updated the code to select the first target env from the `targets` array when multiple entries are present. I’ve also added a corresponding test case to check this behavior.

If you believe selecting another target env would be more appropriate, or if you have any other feedback, please feel free to leave a comment. Otherwise, if everything looks good, I’d appreciate your approval. Thank you!

kuhar commented on mlir/test/Conversion/GPUToSPIRV/lookup-target-env.mlir:
```diff
@@ -0,0 +1,17 @@
+
+module attributes {gpu.container_module} {
+  // CHECK-LABEL: spirv.module @{{.*}} GLSL450
+  gpu.module @kernels [#spirv.target_env<#spirv.vce<v1.0, [Shader], [SPV_KHR_storage_buffer_storage_class]>, #spirv.resource_limits<>>] {
```
Can we have a test with more than one target env attr in the array?

oojahooo replied:
I added it. Thanks!

oojahooo committed:
Select first target env when targets array has more than one target env

oojahooo committed:
Add test case with more than one target env in targets array

kuhar reviewed (approved):
LGTM % one minor request

oojahooo committed:
Add comment explaining the purpose of new test case

oojahooo committed:
Merge branch 'main' into spirv_lookup_target

oojahooo commented:
Hi @kuhar, thank you again for the approval.
If you have a moment, could you also approve the pending workflows when convenient?
I understand it's quite late in your timezone, so no rush at all — just wanted to leave a gentle note here.
Thanks again!

IgWod-IMG commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -48,19 +48,45 @@ struct GPUToSPIRVPass final : impl::ConvertGPUToSPIRVBase<GPUToSPIRVPass> {
+  if (spirv::TargetEnvAttr targetEnvAttr = lookupTargetEnvInTargets(moduleOp))
+    return targetEnvAttr;
+
+  return spirv::lookupTargetEnvOrDefault(moduleOp);
```
Looking at the tests, it looks like they both test the `if` statement above. Is there a test checking this part of the function? If not, do we need a test to check that the function behaves as intended when no attributes are attached?

oojahooo replied:
If I understood correctly, we should test the behavior when the `gpu.module` does not have a `targets` attribute with spirv target env so that the line 78 code is executed.

I believe this case might already be covered by existing tests. For instance, [gpu-to-spirv.mlir](https://github.com/llvm/llvm-project/blob/49d89bc9f43f967051920731f0ec138ca8aaf5ee/mlir/test/Conversion/GPUToSPIRV/gpu-to-spirv.mlir) tests the behavior when no target env is present at all, and [load-store.mlir](https://github.com/llvm/llvm-project/blob/49d89bc9f43f967051920731f0ec138ca8aaf5ee/mlir/test/Conversion/GPUToSPIRV/load-store.mlir) covers the case where the target env is not attached to the`gpu.module`'s `targets` attribute but to the `module`.

However, if a more specific test is needed, such as one where the `gpu.module` lacks the `targets` attribute but has a target env in its attr-dict, or if other additional tests are necessary, I would be happy to add them based on your feedback.

IgWod-IMG replied:
> If I understood correctly, we should test the behavior when the gpu.module does not have a targets attribute with spirv target env so that the line 78 code is executed.

Correct.

> I believe this case might already be covered by existing tests. For instance, [gpu-to-spirv.mlir](https://github.com/llvm/llvm-project/blob/49d89bc9f43f967051920731f0ec138ca8aaf5ee/mlir/test/Conversion/GPUToSPIRV/gpu-to-spirv.mlir) tests the behavior when no target env is present at all, and [load-store.mlir](https://github.com/llvm/llvm-project/blob/49d89bc9f43f967051920731f0ec138ca8aaf5ee/mlir/test/Conversion/GPUToSPIRV/load-store.mlir) covers the case where the target env is not attached to thegpu.module's targets attribute but to the module.

Sounds good, I think we are good.

IgWod-IMG reviewed (approved):
Thanks for addressing my question. LGTM.
